### ISSUE TITLE
Why use RowwiseParallel for nn.Embedding instead of ColwiseParallel?

### ISSUE THREAD

**Issue by corey-lambda** [pytorch/torchtitan#785]
Especially since as configured, the Rowwise parallel shards on the last dimension, which colwise parallel will do automatically:

https://github.com/pytorch/torchtitan/blob/main/torchtitan/parallelisms/parallelize_llama.py#L133

```
    parallelize_module(
        model,
        tp_mesh,
        {
            "tok_embeddings": RowwiseParallel(
                input_layouts=Replicate(),
                output_layouts=Shard(1),
            ),
```

Colwise also makes the logic a bit more clear. Rowwise splits on the token dimension, leading to confusion on how the different shards handle tokens that are not present within their shard. From a bit of debugging it seems like there is a special case for this somewhere deep in pytorch source code, but I could not find it.

With colwise, the embedding weight matrix is split on the model dim dimension, so all shards have all the tokens, just different parts of the model dim. I believe this also would be more performant as no extra work has to be done to transform the output to Shard(1) layout, whereas with rowwise, we do have to do extra sharding work.

Can someone provide some insight?

**Comment by tianyu-l** [comment]
Thanks for the question.

I think both Rowwise and Colwise would work correctly, and it seems to be their performance on memory and computation time should be similar. It is worth noting that vocab dimension is often much larger than the hidden dimension, so technically it has more potential to do Rowwise, although in practice the TP degree is likely not that much.

The difference should be in communication:
- With Rowwise, after the layer the embeddings would be stored in a `_MaskPartial` placement and requires a reduce-scatter to be sharded on the sequence dimension in Sequence Parallel (or all-reduce if vanilla Tensor Parallel)
- With Colwise, the embeddings are sharded on the hidden dim, which requires an all-to-all to be sharded on the sequence dim (or all-gather if vanilla Tensor Parallel).

At the time the code was written, all-to-all for Colwise was not supported yet, and we ended up using Rowwise.

To be honest I haven't tested the perf difference, since all-to-all is supported now. Are you interested in testing and sharing with us, on your hardware and preferred config?

**Comment by xffxff** [comment]
I tested on 8 H100 gpus using llama3_8b.toml, setting the tp size to 8 and the batch size to 4.

Here are the results for memory usage and throughput:
![image](https://github.com/user-attachments/assets/cb84c405-a143-494d-8033-47d1fa9ef35d)

Both Colwise and Rowise had the same max active memory usage, but Colwise had higher max reserved memory usage and it also achieved slightly better throughput.

Colwise Profile
![image](https://github.com/user-attachments/assets/74540ff3-6cba-462f-b296-f8dde978718a)

Rowwise Profile
![image](https://github.com/user-attachments/assets/667ef23d-4569-4f23-b229-25104014d0ee)


It looks like the logic of `aten::embedding` is more complicated with Rowwise than with Colwise. It might because the embeddings are shared on the vocab dim, so the embeddings in local only have a partial embeddings of the whole vocab, that means we can't just do `index_select`, I noticed that `aten:lt` and `aten::ge` were involved, which might be used to check if we can look up the token embedding in local embeddings. Could this be why Colwise is slightly faster?

Also, it looks like the logic of redistribution with Colwise is more complicated than with Rowwise.

**Comment by tianyu-l** [comment]
@xffxff Thanks! This looks interesting.
> It looks like the logic of aten::embedding is more complicated with Rowwise than with Colwise. It might because the embeddings are shared on the vocab dim, so the embeddings in local only have a partial embeddings of the whole vocab, that means we can't just do index_select, I noticed that aten:lt and aten::ge were involved, which might be used to check if we can look up the token embedding in local embeddings.

You are basically right.
> Could this be why Colwise is slightly faster?

That could be the reason, but to verify it from the profile trace, we'll need to look at the GPU part (what you presented is the CPU part, but the throughput should be GPU-bound). Note that the communication all-to-all / reduce-scatter should be exposed, so their difference could also contribute to the throughput difference (again let's compare them in the GPU profile).
> Also, it looks like the logic of redistribution with Colwise is more complicated than with Rowwise.

The redistribution with Colwise is just a `Shard(2)` to `Shard(1)` (hidden dim to sequence dim) redistribution of the underlying DTensor. But it seems incurring multiple `contiguous` calls. cc: @wanchaol

**Comment by xffxff** [comment]
> let's compare them in the GPU profile).

@tianyu-l Thanks for getting back to me!
 
Added some screenshots from the GPU part, some observations:
* The `index_select` kernel took 0.104 ms in the Colwise profile, but it took 0.788 ms in the Rowwise profile
* The `all-to-all` took 7.142 ms in the Colwise profile, but the `reduce scatter` in the Rowwise profile only took 1.456 ms.

I'm not sure how to do more analysis, but hope this helps. I’d be happy to hear any insights you might have!

(I tried uploading the trace files, but they are too big -- Github limits file sizes to 25MB)

Colwise profile with the GPU part:
<img width="1859" alt="image" src="https://github.com/user-attachments/assets/8527d998-88ae-4541-8ed9-b0b87bbd2f72" />
<img width="1824" alt="image" src="https://github.com/user-attachments/assets/246fcc38-09ec-441d-836d-cac9aec55781" />

Rowwise profile with the GPU part:
<img width="1858" alt="image" src="https://github.com/user-attachments/assets/0e2db134-1c25-4a48-9171-d87d3fcf4b0f" />
<img width="1843" alt="image" src="https://github.com/user-attachments/assets/8e601ab1-0623-4655-9962-5c9f0e9c3fc6" />

**Comment by corey-lambda** [comment]
Thanks for all of these wonderful responses!

> It looks like the logic of aten::embedding is more complicated with Rowwise than with Colwise. It might because the embeddings are shared on the vocab dim, so the embeddings in local only have a partial embeddings of the whole vocab, that means we can't just do index_select, I noticed that aten:lt and aten::ge were involved, which might be used to check if we can look up the token embedding in local embeddings.

This is why I ended up raising the issue. I was trying to explain why rowwise worked in a simple way and just kept coming back to colwise being so much easier to explain.

**Comment by tianyu-l** [comment]
Another reason could be: when weight-tying between embedding and output layer is enabled, rowwise for embedding is the same as columnwise for the output layer.
